core-bundle/src/EventListener/DataContainer/PageUrlListener.php

```diff
@@ -0,0 +1,396 @@
+        } catch (DuplicateAliasException $exception) {
+            throw new \RuntimeException($this->translator->trans('ERR.pageUrlPrefix', [$exception->getUrl()], 'contao_default'), $exception->getCode(), $exception);
+        }
+
+        return $value;
+    }
+
+    /**
+     * @Callback(table="tl_page", target="fields.urlSuffix.save")
+     */
+    public function validateUrlSuffix($value, DataContainer $dc)
+    {
+        if ('root' !== $dc->activeRecord->type || $dc->activeRecord->urlSuffix === $value) {
+            return $value;
+        }
+
+        /** @var PageModel|Adapter $pageAdapter */
+        $pageAdapter = $this->framework->getAdapter(PageModel::class);
+
+        $rootPage = $pageAdapter->findByPk($dc->id);
+
+        if (null === $rootPage) {
+            return $value;
+        }
+
+        try {
+            $this->recursiveValidatePages((int) $rootPage->id, $rootPage);
+        } catch (DuplicateAliasException $exception) {
+            throw new \RuntimeException($this->translator->trans('ERR.pageUrlSuffix', [$exception->getUrl()], 'contao_default'), 0, $exception);
+        }
+
+        return $value;
+    }
+
+    public function reset(): void
+    {
+        $this->prefixes = null;
+        $this->suffixes = null;
+    }
+
+    private function purgeSearchIndex(int $pageId): void
+    {
+        $urls = $this->connection->executeQuery(
+            'SELECT url FROM tl_search WHERE pid=:pageId',
+            ['pageId' => $pageId]
+        )->fetchAll(FetchMode::COLUMN);
+
+        foreach ($urls as $url) {
+            $this->searchIndexer->delete(new Document(new Uri($url), 200));
+        }
+    }
+
+    /**
+     * @throws DuplicateAliasException
+     */
+    private function recursiveValidatePages(int $pid, PageModel $rootPage): void
+    {
+        /** @var PageModel|Adapter $pageAdapter */
+        $pageAdapter = $this->framework->getAdapter(PageModel::class);
+
+        $pages = $pageAdapter->findByPid($pid);
+
+        if (null === $pages) {
+            return;
+        }
+
+        /** @var PageModel $page */
+        foreach ($pages as $page) {
+            $this->aliasExists($page->alias, (int) $page->id, $rootPage, true);
+
+            $this->recursiveValidatePages((int) $page->id, $rootPage);
+        }
+    }
+
+    /**
+     * @throws DuplicateAliasException
+     */
+    private function aliasExists(string $currentAlias, int $currentId, PageModel $currentPage, bool $throw = false): bool
+    {
+        $currentPage->loadDetails();
+
+        $currentDomain = $currentPage->domain;
+        $currentPrefix = $currentPage->urlPrefix;
+        $currentSuffix = $currentPage->urlSuffix;
+
+        if ('root' === $currentPage->type) {
+            /** @var Input $input */
+            $input = $this->framework->getAdapter(Input::class);
+
+            // TODO: this won't work in edit-all, in legacy mode or if user does not have access to these fields
```

Comment:
It seems we should address this before we merge?

Reply:
Not sure if we can... That's always the issue with referring to other fields in a save callback 🤷

Reply:
What I mean by "addressing" is that we either bypass the check in these cases or throw an exception in these cases. But we cannot just silently ignore it.